### PR TITLE
avoid overwritting files

### PR DESCRIPTION
**Pull request by Tianhao-Gu**
there shouldn't be any case where we want to overwrite batch scripts if they exist, though I think if they are identical it won't affect anything.

### CODE REVIEW

File: src/loaders/jobs/taskfarmer/task_generator.py [hunk 1]
```diff
@@ -83,6 +87,17 @@ def _pull_image(image_str, job_dir):
...
 
+def _write_to_file(file_path, content, mode="x"):
+    """
+    Writes the specified content to the specified file.
+    """
+    try:
+        with open(file_path, mode) as f:
+            f.write(content)
+    except FileExistsError:
+        print(f"File {file_path} already exists.")
```
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/task_generator.py:98]: This makes me very nervous as it could cause difficult to find bugs if files are overwritten but people aren't carefully checking the logs. I'd prefer to throw the error by default and have a flag to ignore file collisions

**Review by MrCreosote — CHANGES_REQUESTED** [review]

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 1 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py]: 👍

File: src/loaders/jobs/taskfarmer/task_generator.py [hunk 2]
```diff
@@ -83,6 +87,22 @@ def _pull_image(image_str, job_dir):
...
+    If the file already exists, an exception is raised unless ignore_existing is set to True.
+    """
+    try:
+        with open(file_path, mode) as f:
+            f.write(content)
+    except FileExistsError:
+        if ignore_existing:
+            print(f"File {file_path} already exists.")
+        else:
+            raise FileExistsError(f"File {file_path} already exists.")
```
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/task_generator.py:103]:
```suggestion
            raise
```
any reason not to use the same exception?

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 1 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: I meant there should be a cli flag to overwrite existing files

Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py]: that's done by giving the 'w' mode.

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 2 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py]: 👍

Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: :+1:

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 1 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: That's not a CLI flag?

Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: Thinking about this further, IIUC if any of the submission scripts / text files exist, they won't be overwritten by this logic. I guess I don't understand why you wouldn't just generate all new scripts (maybe backing up the old ones if necessary)

Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py]: My aim was to avoid overwriting existing files, particularly in situations where a user runs the script multiple times. For instance, if a user executes the script again without specifying the --force option, it will overwrite all existing files. This may not be problematic if all the arguments remain the same. However, if the user runs the script with a different source_dir argument and without --force, the situation becomes complicated.

Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: I feel like I'm missing something here - IIUC if I run the script twice in a row with different arguments, the script will ignore the 2nd set of arguments because it doesn't overwrite the files. That seems like the exact opposite of the behavior anyone would expect

Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: Maybe we should zoom on Monday and talk through this, this thread is getting long

File: src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py [hunk 3]
```diff
@@ -237,10 +264,6 @@ def _check_preconditions(self, restart_on_demand):
 
             return True
 
-        if latest_task['source_data_dir'] != self.source_data_dir:
```
Comment (Tianhao-Gu) [inline src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py:240]: This seems to be irrelevant at this point. If the previous job failed/canceled/timeout, it's okay to restart the job with a different argument. If the previous job is completed/running/pending, the job won't run unless `force`. If the user force runs the job, it also becomes irrelevant.

File: src/loaders/jobs/taskfarmer/task_generator.py [hunk 4]
```diff
@@ -292,6 +306,7 @@ def main():
     root_dir = args.root_dir
 
     task_mgr = TFTaskManager(kbase_collection, load_ver, tool, source_data_dir, root_dir=root_dir)
+    task_mgr.check_preconditions(args.force)
```
Comment (Tianhao-Gu) [inline src/loaders/jobs/taskfarmer/task_generator.py:309]: I think we should just do the precondition check before writing batch scripts. When the previous job is in complete/running/pending status, we shouldn't update batch scripts. However, when '--force' or the previous job is in failed/canceled/timeout status, we should update the batch script and allow users to resubmit jobs with the same/different arguments.

Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py:309]: after precondition checking is passed, (the previous job is in failed/canceled/timeout status or --force is True), it's safe to overwrite existing batch files.

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 1 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py]: 👍 made a few changes. But sure let's sync up on Monday.

File: src/loaders/jobs/taskfarmer/task_generator.py [hunk 5]
```diff
@@ -83,6 +87,20 @@ def _pull_image(image_str, job_dir):
                              f"Standard error: {std_err.read()}")
 
 
+def _write_to_file(file_path, content, overwrite=True):
```
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/task_generator.py:90]: Looks like none of the calls specified overwrite=False, so I assume that arg can be removed

File: src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py [hunk 6]
```diff
@@ -277,11 +302,7 @@ def _check_preconditions(self, restart_on_demand):
...
-            raise PreconditionError(
-                f'There is a previous run of the same tool and load version with a different source data directory.')
-
-        if latest_task_status in [JobStatus.FAILED, JobStatus.CANCELLED, JobStatus.TIMEOUT]:
+        if latest_task_status in [JobStatus.FAILED, JobStatus.CANCELLED]:
```
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py:305]: Was timeout deleted intentionally?

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 4 above)
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/task_generator.py:309]: If the preconditions check needs to happen every time, just call it from the class constructor

File: src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py (hunk 3 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py:280]: Hmm, what about the case where the job failed on 1 genome but all the others are ok?

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 4 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py]: 👍

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 5 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py]: 👍

File: src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py (hunk 3 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py:280]: we do have a ticket for that situation. But I haven't thought through how we can do this. 
https://kbase-jira.atlassian.net/browse/RE2022-96
I feel this still has nothing to do with `source_data_dir`.

File: src/loaders/jobs/taskfarmer/task_generator.py [hunk 7]
```diff
@@ -305,7 +313,7 @@ def main():
 
     if args.submit_job:
         try:
-            task_mgr.submit_job(restart_on_demand=args.force)
+            task_mgr.submit_job()
         except PreconditionError as e:
```
Comment (MrCreosote) [inline src/loaders/jobs/taskfarmer/task_generator.py:317]: so submit_job can't throw a precondition error as is, but maybe both the constructor and the submit call should check preconditions in case there's a change in between. Either way the I assume the precondition error needs to be caught when calling the constructor

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 1 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: Files are always overwritten if the script proceeds at this point so moot

:+1:

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 5 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: :+1:

File: src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py (hunk 6 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py]: :+1:

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 4 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: :+1:

File: src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py (hunk 3 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py]: added back `source_data_dir` check.

Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py]: :+1:

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 4 above)
Comment (MrCreosote) [reply src/loaders/jobs/taskfarmer/task_generator.py]: :+1:

**Review by MrCreosote — APPROVED** [review]
LGTM

File: src/loaders/jobs/taskfarmer/task_generator.py (hunk 7 above)
Comment (Tianhao-Gu) [reply src/loaders/jobs/taskfarmer/task_generator.py:321]: 👍 updated to catch preconditionError at both places.
